source/language/types.rst

```diff
@@ -12,28 +12,66 @@ Variable identifiers must begin with a letter [A-Za-z], an underscore, a
 percent sign, or an element from the Unicode character categories
 Lu/Ll/Lt/Lm/Lo/NI :cite:`noauthorUnicodeNodate`.
 Continuation characters may contain numbers. Variable identifiers may
-not override a reserved identifier. All qubits are global variables.
+not override a reserved identifier.
+
+All qubits are global variables.
+
 When qubits are declared, their state is initially undefined. Qubits
 cannot be declared within gates or subroutines. This simplifies OpenQASM
 significantly since there is no need for quantum memory management.
-However, it also means that users or compiler have to explicitly manage
-the quantum memory. In addition to being assigned values within a
-program, all of the classical types can be initialized on declaration.
+However, it also means that users or compilers have to explicitly manage
+the quantum memory.
+
+OpenQASM supports all scalar classical types as defined in the C99 Standard:
+http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf
+
+Additionally, OpenQASM 3.0 supports compile-time fixed-size C-style bit
+arrays.  Variable-Length Arrays (VLA's) are not supported in OpenQASM 3.0.
+
+Missing - Needs Clarification:
+------------------------------
+Does an Array automatically decay to the pointer of its first element, as in C?
```

Comment:
What do you mean by "decay"? We don't have pointers in OpenQASM3, so I am not sure what you mean.

Reply:
I wasn't clear with what I mean by "decay", so, here goes:

In C - and C-like derivatives - an array of type T automatically "decays" to a pointer of type T* to said array's first element. This means I can write something like this:

```
int array[10];
int* p = array;
```
Now I can access the elements of array either by index (```array[0], array[1]```, etc), or by pointer arithmetic (```*p, *(p+1), (*p+2)```, etc, etc, etc).

Why is this important: Since OpenQASM 3.0 supports passing arrays as arguments to functions and subroutines, we need to specify **how** these array arguments are passed. If we want to do it "like C does it", then we must recognize that, in C, array function arguments are passed as pointers, not as arrays. I.e. only the pointer to the array's first element is passed as argument. If we don't want to do it "like C does it", then we are implicitly doing it "like FORTRAN does it". Which means, the entire array is passed in as argument.

These two ways of passing arrays as function arguments are mutually exclusive.

Since you indicated above that you would like to have the ability to write ```foo(int[32], a[])``` -- i.e. pass an array of arbitrary size as a function argument -- that implies support for implicit array-to-pointer conversion. I.e. array "decay". I.e. we are implicitly supporting pointers.